1week: implement ToDoList Console Program

leeheefull opened the pull request.

leeheefull committed:
1week: before first code review

leeheefull committed:
1week: study MVC pattern

leeheefull committed:
1week: after first code review

leeheefull committed:
1week: implements ToDoList as MVC pattern

leeheefull committed:
1week: today i learned
- enum
- VO, DAO, DTO, CRUD
- Comparable vs Comparator

leeheefull committed:
1week: after second code review
- interface R 삭제하고 위험성 제거
- ArrayList -> List
- view, controller 나누기
- for each문으로 수정(for문의 index사용 지양)
- final을 이용하여 상수 값을 저장하는 constants class 만들기

leeheefull committed:
1week: upload class diagram

jeonyeonbin commented on 1week/ToDoList/src/ac/hs/se/model/ToDo.java:
```diff
@@ -0,0 +1,66 @@
+		this.id = id;
+		this.date = date;
+		this.content = content;
+		this.status = "TODO";
```
위 status도 static한 변수로 관리되는게 더 좋아보입니다.

leeheefull replied:
수정했습니다.

jeonyeonbin commented on 1week/ToDoList/src/ac/hs/se/controller/ToDoController.java:
```diff
@@ -0,0 +1,89 @@
+				}
+			}
+		} catch (Exception e) {
+			view.showMessage("ERROR");
```
이 프로그램은 실제로 예제로 되어있기에 이렇게 에러만 사용자에게 찍지만,
실제로 어플리케이션을 개발할 시에는 error stack trace도 logging을 해야됩니다.
나중에는 file에 로깅하는것도 한번 고민해보면 좋을것같아요.

leeheefull replied:
네 그 부분에서 고민을 해봤을 때에는 아래와 같이 작성하면 될 것 같다고 생각합니다.
```java
try {
	PrintStream ps = null
	FileOutputStream fos = null
	fos = new FileOutStrearm("error.log", true);
	PrintStream ps = new PrintStream(fos);
	System.setErr(ps);

	// content

} catch (Exception e) {
	System.err.println("-----------------------------------");
	System.err.println("예외발생시간: " + new Date());
	e.printStackTrace(System.err);
	System.err.println("예외 메세지: " + e.getMessage());
	System.err.println("-----------------------------------");
}
```

jeonyeonbin commented on 1week/ToDoList/src/ac/hs/se/module/ToDoModule.java:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	public void selectToDo(int id) {
+		for (int i = 0; i < toDoList.size(); i++) {
```
요소삭제는 잘하셨지만 메소드명이 알맞지 않아보이네요.

leeheefull replied:
요소를 선택한다는 뜻을 가진 selectToDo보다는 요소를 삭제하고 있으니 deleteToDo로 수정하는게 좋을 것 같습니다.
수정했습니다.

jeonyeonbin commented on 1week/ToDoList/src/ac/hs/se/module/ToDoModule.java:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	public List<ToDo> selectToDoList(String status) {
+		List<ToDo> list = new ArrayList<>();
```
제가 짠코드 아니더라도 이런 상태값은 세개이상이 될시에는 enum으로 관리되는게 더좋습니다.

String은 사용자의 자유도를 높이기에 fix된 enum을 parameter로 받는게 더좋아보입니다.

```

public enum TodoType {
	TODO("TODO"),
	DOING("DOING"),
	DONE("DONE");

	private final String type;
	private String getType() {
		return type;
	}
	TodoType(String type) {
		this.type = type;
	}

}


```

enum에 대해서 한번 공부를해보시죠.

leeheefull replied:
enum 학습하였습니다.
enum으로 관리하는게 좋은게 맞는 것 같습니다.
enum으로 수정하였습니다.

jeonyeonbin commented on 1week/ToDoList/src/ac/hs/se/controller/ToDoController.java:
```diff
@@ -0,0 +1,89 @@
+				view.showMenu();
+				String menu = br.readLine();
+				switch (menu) {
+				case Menu.INSERT:
```
static import * 에 대해서도 한번 공부해보시면 좋을것같습니다.

현재 Menu는 전부  다 사용되어 보이네요.

leeheefull replied:
꿀팁 감사합니다.
덕분에 static import와 static import *에 대해서 알게 되었습니다.
수정하였습니다.

jeonyeonbin commented on 1week/ToDoList/todolist_diagram.ucls:
```diff
@@ -0,0 +1,114 @@
+  <class id="1" language="java" name="ac.hs.se.main.ToDoConsole" project="ToDoList" 
+    file="/ToDoList/src/ac/hs/se/main/ToDoConsole.java" binary="false" corner="BOTTOM_RIGHT">    
+    <position height="-1" width="-1" x="581" y="83"/>    
+    <display autosize="true" stereotype="true" package="true" initial-value="false" signature="true" 
```
이런파일들은 올리지않습니다. gitignore java 검색해서 한번 찾아보고 추가해보시죠.

leeheefull replied:
평소에 다른 유저들의 github의 repo를 보면서 어떻게 이렇게 깔끔하게 올릴 수 있는지 궁금했었습니다.
"add로 직접 하는건가..?" 라고 생각했었는데, .gitignore를 사용했다는 것을 이제야 알게 되었습니다.
덕분에 좋은 정보를 알게 되었습니다.
앞으로 .gitignore를 자주 사용하겠습니다. 감사합니다.

jeonyeonbin commented on 1week/ToDoList/src/ac/hs/se/main/ToDoConsole.java:
```diff
@@ -0,0 +1,12 @@
+public class ToDoConsole {
+
+	public static void main(String[] args) throws IOException {
+		new ToDoController().run();
```
IOException이 난다는것은 사용자에게 IOException이 보여질거라고 보이는데요.

그런 메세지를 보여주지않으려면 어떻게해야될까요?

그리고 해당 책임은 run을 시키는 TodoConsole에 두어야될까요? 아니면 TodoController에 두어야될까요?

leeheefull replied:
제가 소스코드를 완벽하게 스캔하지 못한 것 같습니다.
해당 부분은 지우는게 맞는 것 같습니다.

leeheefull committed:
1week: today i learned
- static import *
- .gitignore

leeheefull committed:
1week: last code review
- status를 static 변수로 관리
- method nameing 다시하기
- enum을 이용한 final변수 관리
- static import * 적용하기
- .gitignore 적용하기
